core/src/main/java/io/aiven/kafka/tieredstorage/chunkmanager/ChunkManagerFactory.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright 2023 Aiven Oy
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.aiven.kafka.tieredstorage.chunkmanager;
+
+import java.util.Map;
+import java.util.concurrent.Executor;
+
+import org.apache.kafka.common.Configurable;
+
+import io.aiven.kafka.tieredstorage.ObjectKey;
+import io.aiven.kafka.tieredstorage.security.AesEncryptionProvider;
+import io.aiven.kafka.tieredstorage.storage.ObjectFetcher;
+
+public class ChunkManagerFactory implements Configurable {
+    private ChunkManagerFactoryConfig config;
+
+
+    @Override
+    public void configure(final Map<String, ?> configs) {
+        this.config = new ChunkManagerFactoryConfig(configs);
+    }
+
+    public ChunkManager initChunkManager(final ObjectFetcher fileFetcher,
+                                         final ObjectKey objectKey,
+                                         final AesEncryptionProvider aesEncryptionProvider,
+                                         final Executor executor) {
+        final DefaultChunkManager chunkManager = new DefaultChunkManager(fileFetcher, objectKey, aesEncryptionProvider);
+        if (config.cacheType() == CacheType.NO_CACHE) {
+            return chunkManager;
+        } else {
+            try {
+                final CachingChunkManager<?> cachingChunkManager = (CachingChunkManager<?>) config
+                    .cacheType()
+                    .getChunkManagerClass()
+                    .getDeclaredConstructor(ChunkManager.class, Executor.class)
+                    .newInstance(chunkManager, executor);
+                cachingChunkManager.configure(config.originalsWithPrefix(ChunkManagerFactoryConfig.CHUNK_CACHE_PREFIX));
+                return cachingChunkManager;
+            } catch (final ReflectiveOperationException e) {
+                throw new RuntimeException(e);
+            }
+        }
+    }
+}
```

Comment:
We could refactor this factory to have a clearer hierarchy of `ChunkManager`s: 
 
```suggestion
    public CachingChunkManager<?> initChunkManager(final DefaultChunkManager chunkManager,
                                         final Executor executor) {
        try {
            final CachingChunkManager<?> cachingChunkManager = (CachingChunkManager<?>) config
                .cacheType()
                .getChunkManagerClass()
                .getDeclaredConstructor(DefaultChunkManager.class, Executor.class)
                .newInstance(chunkManager, executor);
            cachingChunkManager.configure(config.originalsWithPrefix(ChunkManagerFactoryConfig.CHUNK_CACHE_PREFIX));
            return cachingChunkManager;
        } catch (final ReflectiveOperationException e) {
            throw new RuntimeException(e);
        }
    }
}
```

Then, move the "no caching" condition to the RSM as always needed, and call caching factory conditionally: 

```java
        final DefaultChunkManager defaultChunkManager = new DefaultChunkManager(fetcher, objectKey, aesEncryptionProvider);
        final ChunkManagerFactoryConfig chunkManagerFactoryConfig = new ChunkManagerFactoryConfig(configs);
        if (chunkManagerFactoryConfig.cacheType() == CacheType.NO_CACHE) {
            chunkManager = defaultChunkManager;
        } else {
            final ChunkManagerFactory chunkManagerFactory = new ChunkManagerFactory();
            chunkManagerFactory.configure(configs);
            chunkManager = chunkManagerFactory.initChunkManager(defaultChunkManager, executor);
        }
```

We can then enforce `CachingChunkManager` requires `DefaultChunkManager` as input instead of any implementation, clarifying the wrapping.

```java
    public CachingChunkManager(final DefaultChunkManager chunkManager, final Executor executor) {
        this.chunkManager = chunkManager;
        this.executor = executor;
    }
```

Finally, `ChunkManager` and `DefaultChunkManager` can be moved back to the root directory, and rename the `chunkmanager` directory back to `cache`. `ChunkManagerFactory` can be also renamed to `CachingChunkManagerFactory` or move this initialization to `RSMConfig`.

Hopefully this leaves all the caching details better contained in its own module

Reply:
I specifically move these things out of RSM since I believe it should not know anything about ChunkManager implementation and it does not matter if caching is enabled or not before I had it implemented exactly like you suggesting. Also regarding package naming, it's not only about caching but also partially chunking. One idea I had was creating the cache subpackage in chunkmanager package maybe and moving the caching part there.

Reply:
Make sense. Agree we can look into `chunkmanager` refactoring later if we cannot think of a better organization atm.